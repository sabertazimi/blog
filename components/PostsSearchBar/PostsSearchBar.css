/**
 * Remove redundant border focus styles:
 * `.ant-select:not(.ant-select-disabled):hover .ant-select-selector`.
 */
.ant-menu .ant-select-selector,
.ant-menu .ant-select:hover .ant-select-selector {
  @apply border-0;
}

/* Remove border radius */
.ant-input-search .ant-input-group .ant-input-affix-wrapper:not(:last-child),
.ant-input-search
  > .ant-input-group
  > .ant-input-group-addon:last-child
  .ant-input-search-button {
  @apply rounded-none;
}
