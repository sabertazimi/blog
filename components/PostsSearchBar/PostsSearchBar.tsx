import AutoComplete from '@components/AutoComplete'
import Input from '@components/Input'
import Link from '@components/Link'
import type { PostMeta } from '@types'
import type { ReactNode } from 'react'
import { useCallback, useState } from 'react'

interface Props {
  posts: PostMeta[]
}

interface Option {
  value: string
  label: ReactNode
}

function PostsSearchBar({ posts }: Props): JSX.Element {
  const [options, setOptions] = useState<Option[]>([])

  const handleSearch = useCallback(
    (value: string) => {
      setOptions(
        value
          ? posts
            .filter(({ title }) =>
              title.toLowerCase().includes(value.toLowerCase()),
            )
            .map(({ slug, title }) => ({
              value: title,
              label: (
                <div>
                  <Link href={`/post/${slug}`}>{title}</Link>
                </div>
              ),
            }))
          : [],
      )
    },
    [posts],
  )

  return (
    <AutoComplete options={options} onSearch={handleSearch}>
      <Input.Search
        allowClear
        enterButton="Search"
        placeholder="Search Posts ..."
      />
    </AutoComplete>
  )
}

export default PostsSearchBar
