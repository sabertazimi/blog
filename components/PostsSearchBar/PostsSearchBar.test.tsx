import mockData from '@mocks/data';
import { fireEvent, render, screen } from '@utils';
import PostsSearchBar from './PostsSearchBar';

describe('PostsSearchBar', () => {
  const mockPosts = mockData.posts;

  beforeEach(() => {
    const { getComputedStyle } = window;
    window.getComputedStyle = element => getComputedStyle(element);
  });

  test('should render correctly (snapshot)', () => {
    const { container } = render(<PostsSearchBar posts={mockPosts} />);

    expect(container).toMatchSnapshot();
  });

  test.each(mockPosts)(
    'should render [%# Basic Notes] options when searching',
    async ({ index, title }) => {
      render(<PostsSearchBar posts={mockPosts} />);
      const input = screen.getByRole('combobox');

      fireEvent.change(input, { target: { value: `${index + 1}` } });

      expect(await screen.findAllByText(title)).toHaveLength(2);

      fireEvent.change(input, { target: { value: '' } });

      expect(await screen.findAllByText(title)).toHaveLength(2);
    }
  );
});
