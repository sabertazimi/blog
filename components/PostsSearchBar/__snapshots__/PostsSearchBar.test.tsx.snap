// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsSearchBar should render correctly (snapshot) 1`] = `
<div>
  <div
    class="ant-select ant-select-auto-complete ant-select-single ant-select-customize-input ant-select-show-search"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <span
          class="ant-input-group-wrapper ant-input-search ant-input-search-with-button ant-select-selection-search-input"
        >
          <span
            class="ant-input-wrapper ant-input-group"
          >
            <span
              class="ant-input-affix-wrapper"
            >
              <input
                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                aria-autocomplete="list"
                aria-controls="rc_select_TEST_OR_SSR_list"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="rc_select_TEST_OR_SSR_list"
                autocomplete="off"
                class="ant-input"
                id="rc_select_TEST_OR_SSR"
                placeholder="Search Posts ..."
                role="combobox"
                type="search"
                value=""
              />
              <span
                class="ant-input-suffix"
              >
                <span
                  class="ant-input-clear-icon ant-input-clear-icon-hidden"
                  role="button"
                  tabindex="-1"
                >
                  <span
                    aria-label="close-circle"
                    class="anticon anticon-close-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </span>
            <span
              class="ant-input-group-addon"
            >
              <button
                class="ant-btn ant-btn-primary ant-input-search-button"
                type="button"
              >
                <span>
                  Search
                </span>
              </button>
            </span>
          </span>
        </span>
      </span>
      <span
        class="ant-select-selection-placeholder"
      />
    </div>
  </div>
</div>
`;
